[5-1] 레스토랑 조회 구현하기 코드 리뷰 요청드립니다.

SeaFalcon opened the pull request:
중간중간 푸쉬하면서 리뷰 요청 드렸어야하는데 완성하고 해야지 미루다가 너무 늦어버렸네요.
TDD 사이클을 적용하려 노렸했지만 아직은 더 숙련도가 필요한 것 같아요.
혹시 커밋 문구나, 작업 단위에 대해서 피드백 해 주실 것이 있다면 부탁드릴게요

중간에 로직을 변경하는 리팩토링을 진행할 때 고생을 많이했네요 ㅠㅠ
처음에 어떻게 설계하고 상태관리를 할 것인지 세밀하게 정하지 않으면 고생하게 된다는 걸 절실하게 깨닫는 한 주 였네요.

커버리지 100% 달성이 생각보다 쉽지 않았습니다. 
특히 Container의 useEffect 부분에서 레스토랑 목록을 가져오는 부분은 
테스트가 어려워서 100% 달성을 위한 테스트 코드를 어거지로 넣었네요.
혹시나 다른 방법이 있으면 의견 부탁드리겠습니다.

SeaFalcon committed:
App.test.jsx 생성

SeaFalcon committed:
App 초기화면 로딩 테스트 추가

SeaFalcon committed:
App 초기화면 로딩 테스트 구현

SeaFalcon committed:
App에서 지역정보 분리 및 AddressList, test 파일 생성

SeaFalcon committed:
AddressList 지역목록 로딩 테스트 추가

SeaFalcon committed:
AddressList 지역목록 로딩 테스트 구현

SeaFalcon committed:
상태를 App에서 분리, addressListContainer 생성

SeaFalcon committed:
AddressListContainer 지역 목록 로딩 테스트 추가

SeaFalcon committed:
AddressListContainer 지역 목록 로딩 테스트 구현

SeaFalcon committed:
mocking을 위해 fixture 폴더 생성 및 데이터 추가

SeaFalcon committed:
AddressList, Container 테스트 mock데이터 사용하도록 리팩토링

SeaFalcon committed:
__mock__ 폴더 생성, App에 useSelector 적용

SeaFalcon committed:
actions, reducer, store 생성 및 초기 값 세팅

SeaFalcon committed:
App에 store 전달

SeaFalcon committed:
reducer loadAddressList 테스트 추가

- fixture mock 데이터 추가 및 로직 수정

SeaFalcon committed:
API Fetch를 위한 api 파일 생성 및 로직 추가

SeaFalcon committed:
store에 react-thunk 적용

SeaFalcon committed:
reducer loadAddressList 테스트 구현

SeaFalcon committed:
reducer 레스토랑 지역 정보 등록  테스트 추가

SeaFalcon committed:
reducer 레스토랑 지역 정보 등록 테스트 구현

SeaFalcon committed:
reducer loadAddressList 테스트 리팩토링

- 비동기 액션에 대한 테스트는 다른 방법이 필요함, 차후에 고려

SeaFalcon committed:
실제 화면에 표시되는지 확인하기 위해  App에 useEffect loadAddressList 액션 로직 추가

SeaFalcon committed:
AddressList 버튼 클릭에 대한 테스트 추가

SeaFalcon committed:
AddressList 버튼 클릭에 대한 테스트 구현

SeaFalcon committed:
AddressListContainer selectAddress (버튼 클릭)  테스트 추가

SeaFalcon committed:
AddressListContainer selectAddress (버튼 클릭) 테스트 구현

- selectAddress 액션 추가
- initialState 프로퍼티 추가
- reducer selectAddress (버튼 클릭) 테스트 추가, 구현

SeaFalcon committed:
AddressListContainer selectAddress (버튼 클릭) 테스트 리팩토링

SeaFalcon committed:
AddressListContainer selectAddress (버튼 클릭) 테스트 리팩토링 2

SeaFalcon committed:
CategoryList.jsx CategoryList.test.jsx  파일 생성

SeaFalcon committed:
CategoryList 카테고리 목록 로딩 테스트 추가

SeaFalcon committed:
CategoryList 카테고리 목록 로딩 테스트 구현

SeaFalcon committed:
CategoryList 카테고리 목록 로딩 테스트 리팩토링

SeaFalcon committed:
CategoryList 버튼 클릭 이벤트 테스트 추가

SeaFalcon committed:
CategoryList 버튼 클릭 이벤트 테스트 구현

SeaFalcon committed:
category 상태를 App에서 분리, CategoryListContainer 생성

SeaFalcon committed:
CategoryListContainer 레스토랑 카테고리 목록 로딩 테스트 추가

SeaFalcon committed:
CategoryListContainer 레스토랑 카테고리 목록 로딩 테스트 구현

SeaFalcon committed:
App.js CatgoryList -> Container로 변경

SeaFalcon committed:
CategoryListContainer selectCategory (버튼 클릭) 테스트 추가

SeaFalcon committed:
actions.js에 selectCategory 추가

SeaFalcon committed:
reducer selectCategory 호출시 상태변화 테스트 추가

SeaFalcon committed:
initialState와 mock State 데이터에 categoryList 프로퍼티 추가

SeaFalcon committed:
reducer selectCategory 호출시 상태변화 테스트 구현

SeaFalcon committed:
CategoryListContainer selectCategory (버튼 클릭) 테스트 구현

SeaFalcon committed:
reducer setCategoryList 호출시 상태변화 테스트 추가

SeaFalcon committed:
actions에 setCategoryList 로직 추가

SeaFalcon committed:
reducer setCategoryList 호출시 상태변화 테스트 구현

SeaFalcon committed:
초기화면 카테고리 리스트 로딩 구현

- api에 fetchCategoryList 로직 추가
- actions에 loadCategoryList 로직 추가
- app의 useEffect에 loadCategory dispatch
- 결론적으로 TDD 적용 못함, 비동기 액션 테스트는 마지막 리팩토링때 도전!

SeaFalcon committed:
ListContainer 테스트 오류 통과 (useSelector 사용)

SeaFalcon committed:
app 레스토랑 목록 표시 테스트 추가

SeaFalcon committed:
app 레스토랑 목록 표시 테스트 구현

SeaFalcon committed:
app에서 레스토랑 목록 분리를 위해 restaurants 파일 생성

SeaFalcon committed:
restaurants  레스토랑 목록 표시 테스트 추가

SeaFalcon committed:
restaurants  레스토랑 목록 표시 테스트 구현

SeaFalcon committed:
restaurants  레스토랑 목록 표시 테스트 리팩토링

SeaFalcon committed:
restaurants  레스토랑 데이터 유무에 따라 목록 표시 테스트 추가, 구현

SeaFalcon committed:
app에서 restaurants 상태관리 분리를 위한 container 파일 생성

SeaFalcon committed:
RestaurantsContainer 레스토랑 목록 표시 테스트 추가

SeaFalcon committed:
RestaurantsContainer 레스토랑 목록 표시 테스트 구현

SeaFalcon committed:
RestaurantsContainer 레스토랑 목록 표시 테스트 리팩토링

- fixture에 테스트시 중복되는 restaurants 데이터 추가

SeaFalcon committed:
restaurantsContainer  레스토랑 데이터 유무에 따라 목록 표시 테스트 추가, 구현

SeaFalcon committed:
App.jsx restaurants > container로 교체

SeaFalcon committed:
api에 fetchRestaurants 로직 추가

SeaFalcon committed:
reducer setRestaurants 액션 상태변화 테스트 추가

SeaFalcon committed:
reducer setRestaurants 액션 상태변화 테스트 구현

SeaFalcon committed:
actions에 loadRestaurants 로직 추가

SeaFalcon committed:
addressListContainer에 useEffect 사용하여, loadRestaurants 호출 => 기능 완성

SeaFalcon committed:
CategoryList 체크표시 활성화 적용, addressListCotainer mock 데이터제거

SeaFalcon committed:
esline error 수정

SeaFalcon committed:
리팩토링 selectedCategory, selectedAddress 자료형 변경

- 레스토랑 목록 api 호출시 객체에서 필요한 정보를 찾아 바로 호출하기 위함

SeaFalcon committed:
각 컴포넌트의 관심사에 맞는 테스트 적용 및 리팩토링

SeaFalcon committed:
reducer 리팩토링

- 커버리지 100% 달성
- if 문 제거

SeaFalcon committed:
API 테스트 적용 및 테스트 커버리지 100% 달성

- redux-mock-store 설치 및 활용
- api.js 리팩토링
- jest에서 fetch 를 테스트하기 위한 node-fetch 설치 및 활용
- 각각 Container의 useEffect의 loadRestaurants를 dispatch하는 부분 테스트를 toBeCalledWith로 처리할 수 없어서 toBeCalled로 테스트 커버리지 달성

SeaFalcon committed:
원격저장소에서 파일명 소문자로 나타나는 현상 수정하기 위한 커밋

SeaFalcon committed:
Rename restaurants.jsx to Restaurants.jsx

소문자 파일명 대문자로 변경

SeaFalcon committed:
Rename restaurants.test.jsx to Restaurants.test.jsx

소문자 파일명 대문자로 변경

wholemann commented on src/AddressList.jsx:
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+
+export default function AddressList() {
```
의미상 Address는 주소이기 때문에 Region이 맞습니다.

wholemann commented on src/reducer.test.js:
```diff
@@ -1,5 +1,15 @@
+    it('레스토랑 지역 정보가 로딩된다.', () => {
+      const state = reducer(initialState, loadAddressList());
 
+      expect(state.addressList).not.toHaveLength(0);
```
지금은 정확한 길이를 표시해주는 게 더 좋아보입니다.

일단 initialState가 fixture로 빠져서 어떤 데이터가 들어가는지 테스트를 읽는 입장에서 알기 어렵습니다.

테스트는 사용 설명서인 동시에 예제입니다.

그리고  fixture는 initialState가 아닙니다.

initialState는 컴포넌트의 초기 상태를 의미하죠. fixture의 뜻을 잘 생각해보면 이건 상태와는 다름을 알 수 있습니다.

fixture는 '고정물'입니다.

이전 강의를 보시면 아시겠지만 fixture는 액션 함수에 전달하는 객체를 fixture로 분리해놨죠.

initialState를 fixture로 빼진 않습니다.

의미적으로도 맞지 않고 일단 initialState를 테스트 외부에서 가져오는 순간 읽기가 어려워 집니다.

이 테스트를 보고 초기 상태가 무엇인지 어떻게 알 수 있을까요?

wholemann commented on src/actions.js:
```diff
@@ -0,0 +1,9 @@
+export function loadAddressList() {
+  return {
+
+  };
```
일단 통과 시킨 상태에서 커밋하는게 좋습니다. 의도를 드러낼 수 있기 때문이죠.

미 구현된 상태로 커밋하는 건 읽는 입장에서도 크게 의미가 없어요.

wholemann commented on src/reducer.test.js:
```diff
@@ -0,0 +1,5 @@
+describe('reducer', () => {
+  it('loadAddressList', () => {
+
+  });
```
테스트를 먼저 작성한 뒤 커밋하시면 좋겠습니다.

빈 테스트는 commit에서 의미를 갖지 못해요.

wholemann commented on src/reducer.js:
```diff
@@ -14,7 +14,11 @@ export default function reducer(state = initialState, action) {
+    const { selectedAddressId } = action.payload;
+
+    const selectedAddress = state.addressList
+      .filter((address) => address.id === selectedAddressId)[0].name;
```
이 부분에서 의도를 더욱 드러내는 방법은 없을까요?

개선할 방법을 고민해보세요. find를 이용한다면 어떨까요?

wholemann commented on src/reducer.test.js:
```diff
@@ -15,9 +15,12 @@ describe('reducer', () => {
     it('레스토랑 지역 정보가 선택된다.', () => {
-      const state = reducer(initialState, selectAddress('서울'));
+      const state = reducer(
+        containAddressState,
+        selectAddress(containAddressState.addressList[0].id),
```
읽는 입장에서 containAddressState가 무엇을 나타내는지

그리고 이곳에서 무슨 일이 벌어지는지 알기 어렵습니다.

이게 다 initialState가 숨어 버려서 그렇습니다.

그리고 굳이 toBe로 name을 비교하기 보단

toEqual을 통한 object 비교가 더 의도를 드러낼 수 있을 듯 하네요.

wholemann reviewed (changes requested):
안녕하세요? 과제 고생 많으셨습니다.

우선 파일 생성이나,  아무런 구현이 없는 빈 테스트를 commit하는 건 크게 의미가 없음을 알려드립니다.

각 commit은 작업의 의미를 나타내야 하는데 무조건 작게 쪼개는 게 능사는 아닙니다.

다소 길더라도 그 변화들이 해당 commit의 의도에 해당된다면 작업 범위와 크기는 큰 문제가 아닙니다.

만약 commit 메시지를 뭘 써야될지 고민된다면 너무 범위가 크거나 의미를 갖기 힘들 정도로

사소한 작업을 했을 확률이 큽니다.

모든 걸 '코드, commit 메시지, PR을 읽는 입장에서 어떻게 느낄지'를 늘 고민하시고 진행하셔야 됩니다.

전반적으로 아쉬운 부분이 반복되어 나타나기 때문에 일단 comment 드린 부분 위주로

모두 수정하신 뒤 다시 리뷰를 진행하면 좋겠습니다.

지금은 initialState를 분리해버려서 테스트를 읽을 때 의도를 알기 어려운 부분이 많네요.

코드리뷰 반영도 너무 잘게 commit을 쪼개실 필요는 없습니다.

여러 파일을 수정하더라도 하나의 문제점에 관련된 변경 사항이라면
commit 메시지를 통해 충분히 이해할 수 있기 때문입니다.

질문은 DM으로 부탁드릴게요:)

wholemann commented on src/CategoryListContainer.jsx:
```diff
@@ -1,18 +1,30 @@
+      }));
+    }
+  }, [selectedCategory]);
+
```
Container에서 굳이 useEffect가 필요 없습니다.

그냥 handleClick에서 모두 dispatch 처리해도 아무런 상관이 없죠.

새로운 레스토랑 목록을 불러와서 상태가 바뀐다면 redux가 이를 알려줄 겁니다.

그러면 이 컴포넌트에선 무슨 일이 벌어질까요?

리액트는 상태를 신경 쓸 필요가 없습니다.

그냥 선택됐다는 상태가 전달된다면 거기에 맞게 V 표시를 해줄 뿐이죠.

지금은 리액트가 카테고리 선택 여부를 신경쓰고 있습니다.

관심사의 분리 측면에서 어떤 게 올바른 방향일까요?

SeaFalcon committed:
address와 관련된 모든 파일명 및 변수명 region으로 변경

SeaFalcon committed:
의도 잘 드러내기

- 배열의 항목 찾기 filter => find로 변경
- fixture에서 initialState 제거
- 개별항목 보다는 Object를 비교하는 tobe => toEqual로 변경
- 의도를 알기 어려운 containAddressState 제거

SeaFalcon committed:
관심사 분리하기

- Container에서 useEffect를 통해 상태체크를 하던 부분 제거
  (action을 통해 dispatch만 하도록 변경)
- api.js fetchRestaurants 예외처리 추가
- reducer에서 loadRestaurants 액션 및 api 호출 테스트하도록 추가

SeaFalcon commented:
피드백 주신 덕분에 많이 깨지고 배워갑니다.

1. 커밋
- 파일 생성이나 구현이 없는 빈 테스트의 커밋은 피해야한다.
- 커밋의 의도를 정했다면 그 범위가 크던 작던 그에 맞게 커밋을 가져가야한다.

2. 의도 나타내기
- 테스트는 의도를 나타내기 쉬워야 한다. 설명서이면서 예제이기 때문,
  특히 초기상태에 대한 부분이 명시되어 있지 않으면 더욱 의도를 파악하기 어렵다.

3. 관심사 분리하기
- 단순하게 리액트는 뷰에 집중, 리덕스는 상태에 집중한다. 명심하자

wholemann commented on src/reducer.js:
```diff
@@ -0,0 +1,40 @@
+      selectedRegion,
+    };
+  },
+  selectCategory: (state, action) => {
```
selectCategory: (state, { payload: { categoryId } }) ...

이러면 밑에서 코드 한 줄 줄일 수 있겠네요ㅎㅎ

SeaFalcon replied:
리펙토링의 세계는 정말 무궁무진하네요 ㅎㅎ

wholemann commented on src/RegionsContainer.jsx:
```diff
@@ -0,0 +1,39 @@
+
+    dispatch(selectRegion(Number(id)));
+    dispatch(loadRestaurants({
+      regionName: regions.find((region) => region.id === Number(id)).name,
+      categoryId: selectedCategory.id,
```
loadRestaurants 쪽에서 getState를 이용한다면

이 부분에서 객체 전달은 필요 없어질 겁니다ㅎㅎ

SeaFalcon replied:
그렇네요! 생각해보니 redux-thunk 를 활용하면 Action creator가 함수를 반환할 때 두번째 매개변수로 getState를 사용할 수 있네요 ㅎㅎ

wholemann commented on src/RegionsContainer.test.jsx:
```diff
@@ -0,0 +1,81 @@
+        <RegionsContainer />
+      ));
+
+      expect(getByText('서울(V)')).toBeInTheDocument();
```
이건 컴포넌트의 관심사가 아닙니다ㅎㅎ V표시는 리덕스에서 보내준 상태를 바탕으로 render 될 뿐이죠.

이곳에서 테스트하진 않습니다.

여기선 클릭이 되는지가 중요할 뿐이죠.

SeaFalcon replied:
결국은 V 표시 또한 상태변화에 따른 변화이기 때문에 
컴포넌트의 입장에서 테스트 할 이유가 없다는 말씀이군요!

wholemann reviewed (commented):
아주 잘하셨네요ㅎㅎ

몇 가지 아쉬운 부분만 comment 남겨놨고

테스트를 좀 더 리팩토링 해보면 좋겠습니다.

'과연 이 컴포넌트가 신경써야 될 부분을 테스트하고 있는지?' 를 점검해보면 좋겠네요.

리팩토링 하시고 싶은 부분은 자유롭게 해보셔도 좋습니다.

SeaFalcon commented:
2차 피드백을 통해 배운 점

1. API 사용법 
- react-thunk 사용시, 두번 째 매개변수 getState 활용
- API 사용전에 github에서 예제 및 사용법 읽어보기

2. 리펙토링은 끝이 없다. 
- 한 번 본 코드도 다시 보자
- 자세히 꼼꼼하게 코드 살펴보기

3. 관심사 분리
- 컴포넌트가 신경써야 될 부분을 테스트하고 있는지, 상태변화를 신경쓰고 있지는 않은지 항상 점검

SeaFalcon committed:
2차 피드백 적용 리펙토링

- react-thunk의 getState 활용하여 loadRestaurants 수정
- 관심사 분리 (UI 에서 상태 변화 감지하여 V 표시 테스트 할 필요 없음)
- 그 외 코드를 줄일 수 있도록 리펙토링 (parameter destructuring)

SeaFalcon committed:
관심사 분리 리펙토링 ( 체크 표시 )

- UI 에서 상태 변화 감지하여 V 표시 테스트 하지 않도록 변경
- 선택된 카테고리와 지역명을 상태관리에서 직접 관리하고 UI 에서는 검사 로직 없이 받은 그대로 표현하도록 변경

SeaFalcon commented:
# 5주차 회고

## 배운 점
### - 자바스크립트 비동기 관련 개념
- Promise
- async - await
- fetch

### - 액션에서 함수를 반환하는 패턴
- Redux에서 비동기 액션 실행하기  (redux-thunk 활용)

### - 테스트 코드는 설명서이자 예제이다.
- 보는 사람이 이해하기 쉽게 의도를 드러내자.

### - 리펙토링은 끝이 없다.
- 완성된 코드도 다시 보자.

### - fixture의 의미
- 고정물, 반복되는 테스트 데이터만 담기

### - 테스트가 외부에 의존해선 안된다.
- API 테스트의 경우, 실제 url에 요청해서 테스트 하지 않는다. (서버가 바뀌면 테스트가 RED로 됨)
- 윤석님 코드 인용 (fetchRegions를 실행하면 지역목록을 받는다 것을 표현한 테스트 코드)
```
describe('fetchRegions', () => {
  global.fetch = jest.fn().mockResolvedValue({
    json() {
      return REGIONS;
    }
  });

  it('returns regions', async () => {
    const regions = await fetchRegions();

    expect(regions).toEqual(REGIONS);
  });
});

// 나 같은 경우는 실제 url을 통해 API 테스트를 진행했고
// fetch 또한 정의되어 있지 않다는 이유로 npm으로 설치해서 강제로 테스트했다.
```

## 좋았던 점
### - 꾸준한 TDD 연습
- 지난 주보다 익숙해짐

### - 다른 사람의 코드 리뷰를 통해 배우는 것들
- sogoagain 님이 mockResolvedValue 사용하는 것을 보고 나도 시도해봐야겠다고 느낌

### - 지난 주 보다 코드숨 학습 시간 증가
- 다음 주는 더 많이!

## 아쉬운 점
### - 4주차때 회고했던 점 지키지 못함

### - 최종결과를 토요일에 첫 푸쉬
- 주중에 과제를 시작했다고 해도 최종결과를 뒤늦게 푸쉬하면 이미 과제에 대한 방향성이 산으로 간 상태
- 중간 중간 리뷰를 받는게 중요할 듯 

### - 의미없는 커밋
- 파일 생성, 빈 테스트 추가와 같이 단순히 커밋의 수만 늘렸던 것 같다

### - 생각없이 진행하는 습관
- 관심사의 분리에 대해 깊게 생각하지 않고 진행한 것 같다. ( 이럴거면 리액트 / 리덕스를 왜 쓰는지 !? 반성하자 )

## 앞으로 할 것
### - 자바스크립트 기본 개념 학습
- 어설프게 알고 있던 개념들 다시 학습하기 ( this, closure, prototype 등등 )

### - 관심사의 분리에 대해 깊게 아주 깊게 생각해보고 진행하기
- 컴포넌트는 보여주는 것만, 리덕스는 상태관리만
- 위 개념을 항상 기억하면서 진행하기

### - 테스트 코드만 읽고도 이해가 갈 수 있게 노력하기
- 의도를 정확하게 전달하기 ( initialState 명시 )
- '과연 이 컴포넌트가 신경써야 될 부분을 테스트하고 있는지?' 확인하기

### - 작업하기 전에 커밋 단위 정하기
- 의미없는 커밋하지 않기
- 커밋의 단위를 정했다면 작업의 양과 상관없이 작업 범위에 맞게 커밋

### - 1일 1푸쉬 (코드리뷰 요청)
- 멘토님 배려하기 + 더 나은 방향성을 피드백 받기

### - 공부한 내용 정리 시작하기
- 깃허브 페이지 활용

### - 해설강의 꼼꼼히 보기
- 아샬님의 좋은 패턴 익히고 수련하고 반복하여 내것으로 만들기

### - 한 주 전보다 발전하기
- 회고에 적었던 내용 지키려 노력하기

wholemann commented on src/Regions.jsx:
```diff
@@ -1,13 +1,12 @@
         <li key={region.id}>
           <button type="button" onClick={onClick} id={region.id}>
             {region.name}
-            {region.name === selectedRegion.name ? '(V)' : ''}
```
이 부분을 삭제하란 말이 아니었고

테스트에서 신경쓸 부분이 아니란 얘기였습니다.

wholemann commented on src/reducer.js:
```diff
@@ -9,27 +9,43 @@ const initialState = {
+      }
+      return {
+        ...region,
+        name: `${region.name}(V)`,
```
이건 반대로 reducer에서 어떻게 그려달라고 정해준 것이 됩니다.

텍스트로 표시하지 않고 아이콘으로 바꾼다면 어떻게 될까요?

리듀서가 '어떻게 그릴지'에 대한 변경 사항에 영향을 받는 게 좋을까요?

wholemann commented on src/CategoryList.jsx:
```diff
@@ -1,13 +1,12 @@
         <li key={category.id}>
           <button type="button" onClick={onClick} id={category.id}>
             {category.name}
-            {category.name === selectedCategory.name ? '(V)' : ''}
```
이 부분은 당연히 리액트 컴포넌트에서 신경써야할 부분이에요. render에 관련된 정책이니까요.

제가 말씀드린 건 테스트로 그걸 검증하는 게 불필요하단 겁니다.

상태에 따라 어떻게 그릴지를 점검하는 건 컴포넌트 테스트에서 집중할 부분이 아니란 겁니다.

wholemann reviewed (changes requested):
아마 피드백 영상을 보셔서 알겠지만

당연히 (V)는 render에 관한 정책이므로 컴포넌트에 유지되어야 합니다.

제가 해당 부분을 컴포넌트 테스트에서 신경쓸 필요가 없다고 말씀드렸는데

오해를 하신 것 같군요ㅠ.ㅠ

기능 구현과 테스트에서 의도를 가지고 검증해야될 부분을 구분할 필요가 있습니다.

영상을 보시고 다시 수정해서 보내주시면 좋겠네요!
